### DISCUSSION TITLE
Any frontend grpc web client code with Bazel example?

### DISCUSSION THREAD

**Discussion by seyoung-song-walmart-com** [grpc/grpc-web#1415] [Q&A - General]
I googled to search for grpc web frontend client code with Bazel example. I've found a few and this looked like the cleanest one to me. https://github.com/stackb/rules_proto/blob/master/example/routeguide/closure/client.js. Then I've found out that closure is in sunset mode. Is there any good alternative?

**Comment by sampajano** [comment]
@seyoung-song-walmart-com Thanks for reaching out!

I took a quick look at the example you linked, it seems to use `grpc_closure_js_library` below:

https://github.com/stackb/rules_proto/blob/5428336c51f18ad628d9c8d7e052a48ec5379590/example/routeguide/BUILD.bazel#L4C78-L4C101

But i'm not familiar with `grpc_closure_js_library` at all (other than it's defined [here](https://github.com/stackb/rules_proto/blob/master/rules/closure/grpc_closure_js_library.bzl) (thanks for @12wrigja for correction)), and from a quick look not sure whether / how it's using the grpc-web library..

By any chance you can directly refer to our npm build in your bazel rule?
https://www.npmjs.com/package/grpc-web

**Reply by 12wrigja** [reply]
The actual starlark definition seems to be here: https://github.com/stackb/rules_proto/blob/5428336c51f18ad628d9c8d7e052a48ec5379590/rules/closure/grpc_closure_js_library.bzl#L5

It is a thin wrapper over the rules_closure closure_js_library rule: https://github.com/stackb/rules_proto/blob/5428336c51f18ad628d9c8d7e052a48ec5379590/rules/closure/grpc_closure_js_library.bzl#L3

I'm not sure what the current maintenance status for rules_closure is - it hasn't ever been maintained by the Closure maintainers.

The Go reference you found is a definition for Gazelle, the Bazel build rule manager.

**Reply by seyoung-song-walmart-com** [reply]
The last release of the rules_closure was last month, so I guess rules_closure repository is actively maintained.
https://github.com/bazelbuild/rules_closure

The wrapper uses this closure_js_library rule.
https://github.com/bazelbuild/rules_closure/blob/master/closure/compiler/closure_js_library.bzl

**Reply by seyoung-song-walmart-com** [reply]
grpc_closure_js_library bazel rule uses this grpc javascript codes https://github.com/stackb/grpc.js/tree/master/js/grpc.

**Reply by sampajano** [reply]
@12wrigja Oh thanks so much for the correction! I had corrected above too!

---

@seyoung-song-walmart-com Thanks so much for the pointers too!

> grpc_closure_js_library bazel rule uses this grpc javascript codes https://github.com/stackb/grpc.js/tree/master/js/grpc.

Oh interesting! Could you point out where in the bazel rule did it mention grpc.js?

It's actually the first time i heard about grpc.js..

I see it is an alternative to this grpc-web project, so I don't think we'd be able to offer too much help as far as this project is concerned.. :)
